spec/rspec/rails/example/controller_example_group_spec.rb

```diff
@@ -19,6 +19,13 @@ def group_for(klass)
       expect(group.included_modules).to include(RSpec::Rails::Matchers::RoutingMatchers)
     end
 
+    it "handles methods invoked via `method_missing` that use keywords" do
+      example = group.new
+      example.define_singleton_method(:has_val?) { |val:| val == 1 }
+      expect(example).to example.have_val(val: 1)
```

Comment:
I'd suggest changing this to:
```ruby
expect { example.have_val(val: 1) }.not_to raise_error
```
as `expect(example)` doesn't participate much and isn't even called when there's this failure. As the exception is coming from `example.have_val` call.
Also, `example.have_val(val: 1)` isn't a valid matcher you can pass to `.to`.

Please let me know what are your thoughts on this.

Reply:
> Also, `example.have_val(val: 1)` isn't a valid matcher you can pass to `.to`

My mistake. It is a matcher, since `have_val` is backed by dynamic matchers.

Reply:
A couple of findings:

1. `super` for `method_missing` is (`actionpack-6.0.4/lib/action_dispatch/testing/assertions/routing.rb`):
```ruby
      # ROUTES TODO: These assertions should really work in an integration context
      def method_missing(selector, *args, &block)
        if defined?(@controller) && @controller && defined?(@routes) && @routes && @routes.named_routes.route_defined?(selector)
          @controller.send(selector, *args, &block)
        else
          super
        end
      end
```

And as of Rails 6.0.4 **it doesn't have** `ruby2_keywords` to it.

2. What triggers the error

```ruby
        def predicate_result
          @predicate_result = actual.__send__(predicate, *@args, &@block)
        end
```

the error happens `after` `method_missing` has called `super`:
```ruby
      example = group.new
      example.define_singleton_method(:has_val?) { |val:| val == 1 }
      dynamic_matcher = example.have_val(val: 1)
      require 'pry'; binding.pry # 
      dynamic_matcher.matches?(example)
```

`matches?`:
```ruby
predicate_accessible? && predicate_matches?
```

`predicate_matches?`:
```ruby
value == predicate_result
```

`predicate_result`:
```ruby
def predicate_result
  @predicate_result = actual.__send__(predicate, *@args, &@block)
end
```

and this `__send__` does trigger.

I barely understand how `ruby2_keywords` on a `method_missing` fixes it.
`predicate` evaluates to `:has_val?`.
```ruby
@args.inspect # => "[{:val=>1}]"
```

Adding `ruby2_keywords :predicate_result` doesn't help.

We do already have `ruby2_keywords` before `method_missing` in dynamic matcher:
```ruby
    def method_missing(method, *args, &block)
      case method.to_s
      when BE_PREDICATE_REGEX
        BuiltIn::BePredicate.new(method, *args, &block)
      when HAS_REGEX
        BuiltIn::Has.new(method, *args, &block)
      else
        super
      end
    end
    ruby2_keywords :method_missing if respond_to?(:ruby2_keywords, true)
```

and `initialize`, too:
```ruby
        def initialize(method_name, *args, &block)
          @method_name, @args, @block = method_name, args, block
        end
        ruby2_keywords :initialize if respond_to?(:ruby2_keywords, true)
```

If I add `ruby2_keywords` to `method_missing` in `controller_example_group.rb`, `@args.inspect` and `predicate` don't change.

`actual` is:
```ruby
> actual # => #<RSpec::ExampleGroups::ApplicationController_2 (no description provided)>
```

and the difference between having and not having `ruby2_keywords before `method_missing` in `ControllerExampleGroup` boils down to:
```ruby
(#<RSpec::Matchers::BuiltIn::Has>)> actual.has_val?(*@args)
ArgumentError: wrong number of arguments (given 1, expected 0; required keyword: val)
```
vs
```ruby
(#<RSpec::Matchers::BuiltIn::Has>)> actual.has_val?(*@args)
=> true
```

while `method_missing` is **not** involved in this call.

I'm out of ideas why this happens, and I tend to think it's some weird side effect, and it's time to summon Jeremy and Benoit for help.

Reply:
Please don't get me wrong, I appreciate that the fix worked, and not opposed to it. My Ruby knowledge of kwargs clearly fails to provide me with understanding why the fix works.

I've expanded the example you came up with, and it works, and prevents from regression, but what it is very suspicious.

If something fishy is going on, and we're addressing the issue incorrectly, or some additional Ruby fix is needed. It's not something that didn't happen in the past.

Reply:
In both cases,
```ruby
> actual.method(:has_val?).parameters
=> [[:keyreq, :val]]
```

Reply:
Bingo!

This is especially confusing (`ruby2_keywords` is present before `method_missing`):
```ruby
    57: def predicate_result
    58:   require 'pry'; binding.pry
 => 59:   @predicate_result = actual.__send__(predicate, *@args, &@block)
    60: end

> @args
=> [{:val=>1}]
> actual.has_val?(*@args)
=> true

> a = [{val: 1}]
=> [{:val=>1}]
> actual.has_val?(*a)
ArgumentError: wrong number of arguments (given 1, expected 0; required keyword: val)
from /Users/pirj/source/rspec-dev/repos/rspec-rails/spec/rspec/rails/example/controller_example_group_spec.rb:24:in `block (3 levels) in <module:Rails>'
```

I vaguely recall that an Array has some internal meta-attribute to it that indicates that it might be passed and expanded to keywords.
And this where it gets ridiculously hard to understand.

So, apparently, when we initially pass `val: 1` to `have_value`, it's tagged as such. We then save it to `@args` in dynamic matchers. And then, when dynamic matchers call the real predicate, `*@args` expand it to kwargs properly.

Without `ruby2_keywords`, this tagging doesn't happen.

And, obviously, to write a regression-preventing spec around it that is not perplexing at the same time.

@eregon Can you please confirm this theory? Do I understand it correctly that there is a special tag that allows an Array to be expanded to kwargs?

Reply:
See https://docs.ruby-lang.org/en/master/Module.html#method-i-ruby2_keywords and https://docs.ruby-lang.org/en/master/Hash.html#method-c-ruby2_keywords_hash-3F
The flag is on the Hash itself, and it only matters for calls with a splat, where the flagged Hash is the last element.

Reply:
Thank you!
TIL `ruby2_keywords_hash?` 👍

Reply:
> `predicate_result`:
> ```ruby
> def predicate_result
>   @predicate_result = actual.__send__(predicate, *@args, &@block)
> end
> ```
>
> and this `__send__` does trigger.
>
> I barely understand how `ruby2_keywords` on a `method_missing` fixes it.
> `predicate` evaluates to `:has_val?`.
> ```ruby
> @args.inspect # => "[{:val=>1}]"
> ```
>
> Adding `ruby2_keywords :predicate_result` doesn't help.
>
> -- https://github.com/rspec/rspec-rails/pull/2509#discussion_r678474813

Yeah, it needs to called where the hash is created, because it's not about how our code calls the method or passes the arguments (with the `ruby2_keywords` solution, anyway).

It looks like you figured it out, but I wrote up this example before I read that, so I'll post it anyway. This is what finally made it click for me, b/c my natural implementation expectations were incorrect and making it harder for me to reason about:

```ruby
def call_this(val:)
  "val is #{val}"
end

def record_args(*args)
  args
end

args = record_args val: 1 # => [{:val=>1}]

# It's a normal hash, Ruby won't pass it into the keywords slot
Hash.ruby2_keywords_hash? args.last # => false
call_this *args rescue $!           # => #<ArgumentError: wrong number of arguments (given 1, expected 0; required keyword: val)>

# We can explicitly turn it into a keywords hash
keywords = args[-1]
Hash.ruby2_keywords_hash? keywords           # => false
keywords = Hash.ruby2_keywords_hash keywords # => {:val=>1}
Hash.ruby2_keywords_hash? keywords           # => true

# Which will indicate to Ruby that it should pass it into the keywords slot
call_this *args rescue $! # => #<ArgumentError: wrong number of arguments (given 1, expected 0; required keyword: val)>
args[-1] = keywords
call_this *args rescue $! # => "val is 1"

# `ruby2_keywords` modifies the method to implicitly do what we did above
Hash.ruby2_keywords_hash? record_args(val: 1).last # => false
ruby2_keywords :record_args
Hash.ruby2_keywords_hash? record_args(val: 1).last # => true

# So now any method anywhere can splat that array into args
call_this *record_args(val: 1) # => "val is 1"
```

Reply:
Playing with it a bit more, I figured out how they do it. It's a flag on the hash. https://gist.github.com/JoshCheek/43b15b9123e337ee6a118cb6ff4c7e7a

Reply:
Bitwise precision aim at the root cause! 👏